File: java-kotlin-project/app/src/test/java/example/java/basic/NullSafetyTest.java
package example.java.basic;

import example.java.dataclass.Person;
import example.java.dataclass.User;
import org.junit.jupiter.api.Test;
import static org.junit.jupiter.api.Assertions.*;

import java.util.Optional;

public class NullSafetyTest {

    private final NullSafety nullSafety = new NullSafety();

    @Test
    public void getUpperCaseNameTest() {
        Person person = new Person("jake", 33, "jake@gmail.com");

        // 정상적인 Person 객체를 전달한 경우
        String result = nullSafety.getUpperCaseName(person);
        assertEquals("JAKE", result);

        // null을 전달한 경우
        String nullResult = nullSafety.getUpperCaseName(null);
        assertEquals("UNKNOWN", nullResult);
    }

    @Test
    public void findUserTest() {
        // 존재하는 사용자 검색
        Optional<User> result = nullSafety.findUser("john");
        assertNotNull(result);
        assertTrue(result.isPresent());

        // 존재하지 않는 사용자 검색
        Optional<User> notFoundResult = nullSafety.findUser("who?");
        assertNotNull(notFoundResult);
        assertFalse(notFoundResult.isPresent());

        // null userId 전달
        Optional<User> nullResult = nullSafety.findUser(null);
        assertNotNull(nullResult);
        assertFalse(nullResult.isPresent());
    }

    @Test
    public void getUserEmailDomainTest() {
        // 정상적인 userId 전달
        String result = nullSafety.getUserEmailDomain("jane");
        assertNotNull(result);

        // null userId 전달
        String nullResult = nullSafety.getUserEmailDomain(null);
        assertNotNull(nullResult);
    }
}

File: java-kotlin-project/app/src/test/kotlin/example/kotlin/basic/NullSafetyTest.kt
package example.kotlin.basic

import example.kotlin.dataclass.Person
import example.kotlin.dataclass.User
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.Assertions.*
import org.junit.jupiter.api.assertThrows

class NullSafetyTest {

    private val nullSafety = NullSafety()

    @Test
    fun getUpperCaseNameTest() {
        val person = Person("jake", 33, "jake@gmail.com")
        val upperName = nullSafety.getUpperCaseName(person);
        assertEquals("JAKE", upperName);

        val upperNull = nullSafety.getUpperCaseName(null)
        assertEquals("UNKNOWN", upperNull);
    }

    @Test
    fun getDisplayNameTest() {
        val user = User("jane", "jane@gmail.com", 25, false)
        val userName = nullSafety.getUserDisplayName(user)
        assertEquals("jane", userName)

        val userNoName = nullSafety.getUserDisplayName(null)
        assertEquals("Anonymous", userNoName)
    }

    @Test
    fun getUserEmailDomainTest() {
        val user = User("jin", "jin@gmail.com")
        val userEmailDomain = nullSafety.getUserEmailDomain(user)
        assertEquals("gmail.com", userEmailDomain)

        val userNoMail = User("jin")
        val userEmailNull = nullSafety.getUserEmailDomain(userNoMail)
        assertEquals("unknown", userEmailNull)

        val userNull = nullSafety.getUserEmailDomain(null)
        assertEquals("unknown", userNull)
    }

    @Test
    fun processUserTest() {
        val user = User("jane", "jane@gmail.com", 25, false)

        // processUser는 void 메서드이므로 예외가 발생하지 않는지 확인
        assertDoesNotThrow {
            nullSafety.processUser(user)
        }

        // null 사용자도 예외가 발생하지 않는지 확인
        assertDoesNotThrow {
            nullSafety.processUser(null)
        }
    }

    @Test
    fun riskyFunctionTest() {
        // 정상적인 문자열 전달
        assertDoesNotThrow {
            nullSafety.riskyFunction("test")
        }

        // null 전달 시 NullPointerException 발생 확인
        assertThrows<NullPointerException> {
            nullSafety.riskyFunction(null)
        }
    }

    @Test
    fun smartCastExampleTest() {
        // 정상적인 문자열 전달
        assertDoesNotThrow {
            nullSafety.smartCastExample("test")
        }

        // null 전달해도 예외 발생하지 않음 (if문에서 체크하므로)
        assertDoesNotThrow {
            nullSafety.smartCastExample(null)
        }
    }
}